# Исследование надежности заемщиков


#### Описание проекта
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

#### Описание данных
children — количество детей в семье  
days_employed — общий трудовой стаж в днях  
dob_years — возраст клиента в годах  
education — уровень образования клиента  
education_id — идентификатор уровня образования  
family_status — семейное положение  
family_status_id — идентификатор семейного положения  
gender — пол клиента  
income_type — тип занятости  
debt — имел ли задолженность по возврату кредитов  
total_income — ежемесячный доход  
purpose — цель получения кредита

## Общая информация о данных

In [2]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

In [7]:
data['days_employed'] = data['days_employed'].abs()

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставляем их как есть. Тем более этот столбец не понадобится для исследования.

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

In [15]:
data['education'] = data['education'].str.lower()

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Исследуем данные

#### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [24]:
data_kids = data.pivot_table(index = ['children'], values = 'debt', aggfunc = ['mean', 'sum', 'count'])
display(data_kids)

mean   sum  count
              debt  debt   debt
children                       
0         0.075438  1063  14091
1         0.092346   444   4808
2         0.094542   194   2052
3         0.081818    27    330
4         0.097561     4     41
5         0.000000     0      9

**Вывод:** 
в таблице:
mean - процент не вернувших кредит в срок от общего числа семей с указанным количеством детей, где 
0 - вернули в срок, 1 - не вернули
sum - количество семей, не вернувших кредит
count - общее количество семей с указанным количеством детей

Из таблицы видно, что общее количество семей с разным количеством детей сильно разнится, семей без детей, взявших кредит, 14091, а семей с 3,4, 5 детьми немного. То есть выборки получаются не корректными для сравнения. Тем не менее, если не брать в расчет многодетные семьи, как сильно выбивающиеся из общей картины, то можно увидеть, что процент семей без детей возвращающих кредит в срок, выше, чем у семей с детьми.

#### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
data_status = data.pivot_table(index = ['family_status'], values = 'debt', aggfunc = ['mean', 'sum', 'count'])
display(data_status)


mean  sum  count
                           debt debt   debt
family_status                              
Не женат / не замужем  0.097639  273   2796
в разводе              0.070648   84   1189
вдовец / вдова         0.066246   63    951
гражданский брак       0.093130  385   4134
женат / замужем        0.075606  927  12261

**Вывод:** 
в таблице:
mean - процент не вернувших кредит в срок от общего числа людей с указанным семейным положением, где 
0 - вернули в срок, 1 - не вернули
sum - количество людей, не вернувших кредит
count - общее количество людей с указанным семейным положением

В данной таблице также есть сильно выбивающая группа - это те, чье семейное положение "женат / замужем". Остальные группы по численности более менее похожи друг на друга. Те, у кого семейное положение "Не женат / не замужем" и "гражданский брак" не возвращают кредит в срок практически в 10% случаев, в отличие от тех, у кого семейное положение "в разводе" и "вдовец / вдова", среди которых процент не вернувших кредит в срок около 6-7%.

#### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
# Для сравнения возьмем столбец total_income_category, в котором в коде выше (2.6) разбивали уровень дохода по категориям
data_income = data.pivot_table(index = ['total_income_category'], values = 'debt', aggfunc = ['mean', 'sum', 'count'])
display(data_income)


mean   sum  count
                           debt  debt   debt
total_income_category                       
A                      0.080000     2     25
B                      0.070602   354   5014
C                      0.084982  1353  15921
D                      0.060172    21    349
E                      0.090909     2     22

**Вывод:** 
в таблице:
mean - процент не вернувших кредит в срок от общего числа людей с указанным уровнем дохода, где 
0 - вернули в срок, 1 - не вернули
sum - количество людей, не вернувших кредит
count - общее количество людей с указанным уровнем дохода

Исключив маленькие выборки, оставляем для сравнения 2 категории - группа с доходом В - 200001–1000000 
и группа с доходом С - 50001–200000. В группе В процент не вернувших вовремя кредит составляет 7%, в то 
время как в группе С - 8,5%. Можем сделать вывод, что чем выше доход, тем выше вероятность возврата 
кредита в срок.

#### Как разные цели кредита влияют на его возврат в срок?

In [29]:
# Для сравнения возьмем столбец purpose_category, в котором в коде выше (2.6) разбивали цели кредита по категориям
data_purpose = data.pivot_table(index = ['purpose_category'], values = 'debt', aggfunc = ['mean', 'sum', 'count'])
display(data_purpose)

mean  sum  count
                              debt debt   debt
purpose_category                              
операции с автомобилем    0.093480  400   4279
операции с недвижимостью  0.072551  780  10751
получение образования     0.092528  369   3988
проведение свадьбы        0.079118  183   2313

**Вывод:** 
в таблице:
mean - процент не вернувших кредит в срок от общего числа людей с указанной целью кредита, где 
0 - вернули в срок, 1 - не вернули
sum - количество людей, не вернувших кредит
count - общее количество людей с указанной целью кредита

В таблице есть сильно выбивающая по численности группа - это те, кто брал кредит на операции с недвижимостью. Остальные группы по численности более менее похожи друг на друга. Процент не вернувших вовремя кредит не сильно зависит от цели кредита, он колеблется примерно от 7,9% до 9,3%. 

Стоит отметить, что среди тех, кто брал кредит на операции с недвижимостью и на проведение свадьбы
процент не вернувших вовремя кредит, ниже, чем тех, кто брал кредиты на авто и образование - примерно 7% 
против 9%.

#### Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
В исходных данных пропуски появились в стобцах days_employed и total_income. Это могли быть случайные пропуски, чтобы не ставить 0 или в силу человеческого фактора. Могли быть неслучайные пропуски, кто-то намеренно не хотел отвечать на эти вопросы. 

#### Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 
Мы не можем увидеть, нет ли нестандартных значений в огромном массиве данных количественных переменных. Если таковые имеются, то смотреть среднее значение здесь будет уже неправильно, оно будет отклоняться в сторону нестандартных значений. Поэтому медиана при работе с количественными переменными лучше, она находится точно посередине всех значений. 

### Общий вывод.

<b><p> 1. Семьи без детей отдают возвращают кредиты в сроки "активнее", чем семьи с детьми - процент должников среди семей без детей 7,5%, в то время как среди семей с детьми эта цифра около 9%.</p>
<p>2. Те, у кого семейное положение "Не женат / не замужем" и "гражданский брак" не возвращают кредит в срок практически в 10% случаев, в отличие от людей, когда-либо вступавших в брак, среди которых процент не вернувших кредит в срок около 6-7%.</p>
<p>3. Чем выше доход, тем выше вероятность возврата кредита в срок.</p> 
<p>4. Цели кредита не оказывают существенного влияния на возврат кредитов. Тем не менее, те, кто берет кредиты на операции с недвижимостью и на свадьбу, возвращают
кредит в срок, чаще, чем те, кто берет их на операции с автомобилем и на образование.</p>
</b> 
